src/metabase/lib/metadata.cljc

```diff
@@ -0,0 +1,137 @@
+(ns metabase.lib.metadata
+  "TODO -- should this just be part of [[metabase.lib.interface]]?"
+  (:require
+   [metabase.lib.dispatch :as lib.dispatch]
+   [metabase.lib.schema.common :as lib.schema.common]
+   [metabase.lib.schema.id :as lib.schema.id]
+   [metabase.shared.util.i18n :as i18n]
+   [metabase.util.malli :as mu]))
+
+(def ^:private FieldMetadata
+  [:map
+   [:lib/type [:= :metadata/field]]
+   [:id {:optional true} ::lib.schema.id/field]
+   [:name ::lib.schema.common/non-blank-string]])
+
+(def ^:private TableMetadata
+  [:map
+   [:lib/type [:= :metadata/table]]
+   [:id ::lib.schema.id/table]
```

Comment:
Maybe `FieldMetadata` isn't the right term here, since this really is referring to the columns that come back in `data.cols` and/or `data.results_metadata.columns` in a QP response **or** to metadata derived from the `Field`s in the application Database. e.g. this FieldMetadata might be the results of a native query or an aggregation something, in which case it wouldn't have have an ID -- what's the Field ID of `count` in `SELECT count(*) AS "count"`? 

I'll make a note and we can figure out how to make this less confusing as a follow-on

OTOH a capital-T Table always has an ID.

Reply:
I'll change the name to `ColumnMetadata` and add more extensive documentation